crates/faucet/src/types.rs

```diff
@@ -64,36 +17,45 @@ impl std::fmt::Display for AssetAmount {
 
 impl AssetAmount {
     /// The absolute maximum asset amount allowed by the network.
-    ///
-    /// An [`AssetAmount`] is further restricted to the values allowed by
-    /// [`AssetOptions`].
     pub const MAX: u64 = FungibleAsset::MAX_AMOUNT;
 
-    pub fn inner(self) -> u64 {
+    /// Creates an [`AssetAmount`] from a base unit amount.
+    ///
+    /// Returns an error if the amount is greater than the maximum allowed amount.
+    pub fn new(base_units: u64) -> Result<Self, AssetAmountError> {
+        if base_units > Self::MAX {
+            return Err(AssetAmountError::AssetAmountTooBig(base_units));
+        }
+
+        Ok(Self(base_units))
+    }
+
+    /// Returns the asset amount in base units.
+    pub fn base_units(&self) -> u64 {
         self.0
     }
-}
 
-impl<'de> Deserialize<'de> for AssetOptions {
-    fn deserialize<D>(de: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        let options = Vec::<u64>::deserialize(de)?;
+    /// Adds another [`AssetAmount`] to the current one and returns the result if it is valid.
+    pub fn checked_add(self, other: Self) -> Option<Self> {
+        Self::new(self.0.checked_add(other.0)?).ok()
+    }
 
-        AssetOptions::new(options).map_err(|invalid_option| {
-            de::Error::invalid_value(
-                de::Unexpected::Unsigned(invalid_option),
-                &format!(
-                    "Maximum fungible asset value allowed by the network is {}",
-                    AssetAmount::MAX
-                )
-                .as_str(),
-            )
-        })
+    /// Subtracts another [`AssetAmount`] from the current one and returns the result if it is
+    /// valid.
+    pub fn checked_sub(self, other: Self) -> Option<Self> {
+        Self::new(self.0.checked_sub(other.0)?).ok()
```

Comment:
This naming is fine but I also don't mind just `add()` and `sub()` because we don't have the unchecked versions.